File: 2020/bj1102.py
import sys
sys.setrecursionlimit(1000000000) 

def ps(station,store,n,remain,answer): # remain : 앞으로 더 켜야할 발전소 수
    minimum = float('inf')
    if(remain == 0):
        return 0

    if(answer[store][remain] != -1):
        return answer[store][remain]
    for i in range(0,n):
        if(store & 1 << (n-1-i)): # 이미 갔던곳
            pass
        else: # 안간곳 
            tmp = find(station,store,n,i) + ps(station,store+(1 << (n-i-1)),n,remain-1,answer)
            if(minimum > tmp):
                minimum = tmp
    answer[store][remain] = minimum
    return minimum

def find(station,store,n,target):
    minimum = float('inf')
    for i in range(0,n):
        if(store & 1 << (n-1-i)): # 이미 갔던곳
            tmp = station[i][target]
            if(minimum > tmp):
                minimum = tmp
    return minimum


n = int(input(""))
station = [[0]*n for i in range(n)]
for i in range(0,n):
    station[i] = list(map(int, input().split()))
state = input("")
p = int(input(""))

total = 0
possible = 0
for i in range(0,len(state)):
    if(state[i] == "Y"):
        total = total + (1 << (len(state)-i-1))
        possible = possible + 1

store = total
answer = [[-1]*n for i in range(66000)]


if(possible == 0):
    print(-1)
elif(possible >= p):
    print(0)
else:
    print(ps(station,store,n,p-possible,answer))

# 5
# 0 9 6 6 6
# 6 0 6 6 1
# 6 20 0 6 6
# 6 30 6 0 6
# 6 50 6 6 0
# NNYYY
# 4

# 16
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
# YYYYYYYYYYYYYYYYY
# 16


# 6
# 0 23 43 14 12 11
# 34 0 43 42 14 34
# 14 29 0 23 28 14
# 19 22 14 0 19 11
# 50 23 13 23 0 45
# 45 24 13 14 11 0
# YNYNNN
# 4

File: 2020/bj1102.c
// 미친문제였다;
// 먼저 16개의 발전소들의 상태를 저장하기 위해서는 bitmask를 사용해야하고
// 이미 방문해본 발전소 상태(bitmask 현황)을 dp에 저장해서 중복 방문을 막고
// 현재까지 최소치를 저장해서 진입하다가 최소치보다 커지면 바로 재귀종료.

#include <stdio.h>
#include <string.h>
#define INF 987654321;

int cost[17][17];
int n,k;
char state[20];
int st;
int dp[70000];
int minimum = INF;

int dev(int now, int count, int total);
int min(int x, int y);

// 전부 N이 들어올때 p=0 -> 0
// 전부 N이 들어올때 p != 0 -> -1
// 처음에 Y개수보다 P가 작거나 같을때 -> 0

int main(){
    int i,j,count;
    scanf("%d",&n);
    for(i=0;i<n;i++){
        for(j=0;j<n;j++){
            scanf("%d",&cost[i][j]);
        }
    }
    scanf("%s",&state);
    scanf("%d",&k);
    i = strlen(state);
    j = 0;
    // printf("i : %d\n",i);
    st = 0;
    count = 0;
    while(i >= 0){
        if(state[i] == 'Y'){
            st = st + (1<<(j-1));
            count++;
        }
        i--;
        j++;
    }
    if(st == 0){
        if(k == 0){
            printf("0");
        }
        else{
            printf("-1");
        }
    }
    else{
        printf("%d",dev(st,count,0));
    }
    // printf("st : %d\n",st);


}

int dev(int now, int count, int total){ // now : 현재발전소 상태 비트, count : 켜져있는 개수
    // input : dev(st,count);
    int m = INF;
    int i,j;
    if(count >= k){
        if(minimum > total){
            minimum = total;
        }
        return 0;
    }
    if(dp[now] != 0){
        // printf("used");
        return dp[now];
    }
    if(total >= minimum){
        return INF;
    }
    for(i=0;i<n;i++){
        if((now & (1<<i)) > 0){ // 현재발전소 켜져있음
            for(j=0;j<n;j++){
                if((now & (1<<j)) == 0){ // 현재발전소 꺼져있음
                    m =  min(m,dev(now +(1<<j),count+1,total+cost[n-i-1][n-j-1])+cost[n-i-1][n-j-1]);
                }
            }
        }
    }
    dp[now] = m;
    return m;
}

int min(int x, int y){
    if(x < y){
        return x;
    }
    else{
        return y;
    }
}
